### PR TITLE
Add Azure Functions User Guide Page

### PR DESCRIPTION
**Pull request by lafernando**
## Purpose
$subject.

## Check List

- [x] **Page Addition**
  - [ ] Add `permalink` to pages

### CODE REVIEW

File: learn/deployment/azurefunctions.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+---
+layout: ballerina-left-nav-pages
+title: Azure Functions
+description: See how the Ballerina deployment in Azure Functions works
+keywords: ballerina, programming language, serverless, cloud, Azure, Functions
+permalink: /learn/deployment/azurefunctions/
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:6]:
```suggestion
permalink: /learn/deployment/azure/
```

File: learn/deployment/azurefunctions.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+active: azurefunctions
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:7]:
```suggestion
active: azure
```

File: learn/deployment/azurefunctions.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+redirect_from:
+  - /learn/deployment/azurefunctions
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:9]:
```suggestion
  - /learn/deployment/azure
```

File: learn/deployment/azurefunctions.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+---
+
+# Azure Functions
+
+The Azure Functions extension provides the functionality to expose a Ballerina function as an Azure Function. This is done by importing the `ballerinax/azure.functions` module and simply annotating the Ballerina function with the `functions:Function` annotation.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:14]:
```suggestion
The Azure Functions extension provides the functionality to expose a Ballerina function as an Azure function. This is done by importing the `ballerinax/azure.functions` module and simply annotating the Ballerina function with the `functions:Function` annotation. 
```

File: learn/deployment/azurefunctions.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+active: azurefunctions
+redirect_from:
+  - /learn/deployment/azurefunctions
+---
+
+# Azure Functions
+
+The Azure Functions extension provides the functionality to expose a Ballerina function as an Azure Function. This is done by importing the `ballerinax/azure.functions` module and simply annotating the Ballerina function with the `functions:Function` annotation. 
+
+An Azure Function consists of a trigger and optional bindings. A trigger defines how a function is invoked. A binding is an approach where we can declaratively connect other resources to the function. There are *input* and *output* bindings. An input binding is a source of data into the function, where an output binding allows us to output data from the function out to an external resource. For more information, check [Azure Functions triggers and bindings concepts](https://docs.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings).
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:16]:
```suggestion
An Azure Function consists of a trigger and optional bindings. A trigger defines how a function is invoked. A binding is an approach in which you can declaratively connect other resources to the function. There are *input* and *output* bindings. An input binding is a source of data into the function. An output binding allows to output data from the function out to an external resource. For more information, go to [Azure Functions triggers and bindings concepts](https://docs.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings).
```

File: learn/deployment/azurefunctions.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+The following Azure Functions triggers and bindings are currently supported in Ballerina:
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:18]:
```suggestion
The following Azure functions triggers and bindings are currently supported in Ballerina:
```

File: learn/deployment/azurefunctions.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+- HTTP trigger, input binding, and output binding
+- Queue trigger and output binding
+- Blob trigger, input binding, and output binding
+- Twilio SMS output binding
+- CosmosDB trigger, input binding, and output binding
+- Timer trigger
+
+The following Ballerina code gives an example of using an HTTP trigger to invoke the function, and a queue output binding to write an entry to a queue, and also an HTTP output binding to respond back to the caller with a message.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:26]:
```suggestion
The following Ballerina code gives an example of using an HTTP trigger to invoke the function, a queue output binding to write an entry to a queue, and also an HTTP output binding to respond back to the caller with a message. 
```

File: learn/deployment/azurefunctions.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+```
+
+The first parameter with the [Context](https://ballerina.io/learn/api-docs/ballerinax/azure-functions.html#Context) object contains the information and operations related to the current function execution in Azure Functions such as the execution metadata and logging actions to be used by the function. This parameter is optional, and can exist at any position in the parameter list of the function. The second parameter with the `HTTPTrigger` annotation signals that this function is going to have an HTTP trigger, and that its details should be stored in the given `HTTPRequest` value. Then we also declares that we will be having a queue output binding by the usage of the `QueueOutput` annotation, with a string result, by defining a `StringOutputBinding` parameter. Also, we declare an HTTP output binding by annotating the return type `HTTPBinding` with the `HTTPOutput` annotation. This HTTP output binding can also be given by defining as a parameter as well, with the same annotation. In this manner, we can mix and match any combination of triggers, input/output bindings, and with or without the execution context object in defining an Azure Function.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:41]:
```suggestion
The first parameter with the [Context](https://ballerina.io/learn/api-docs/ballerinax/azure-functions.html#Context) object contains the information and operations related to the current function execution in Azure Functions such as the execution metadata and logging actions to be used by the function. This parameter is optional and can exist at any position in the parameter list of the function. The second parameter with the `HTTPTrigger` annotation signals that this function is going to have an HTTP trigger and that its details should be stored in the given `HTTPRequest` value. Then, you also declare that you will be having a queue output binding by the usage of the `QueueOutput` annotation with a string result  by defining a `StringOutputBinding` parameter. Also, you declare an HTTP output binding by annotating the return type `HTTPBinding` with the `HTTPOutput` annotation. This HTTP output binding can also be given by defining as a parameter as well with the same annotation. In this manner, you can mix and match any combination of triggers, input/output bindings with or without the execution context object when defining an Azure Function. 
```

File: learn/deployment/azurefunctions.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+In order to deploy a Ballerina function in Azure Functions, the following prerequisites must be met.
+
+* Create an Azure [Function App](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal) with the given resource group with the following requirements:
+   - Runtime stack - `Java`
+   - Hosting operating system - `Windows` (default; Linux is not supported in Azure for custom handlers at the moment)
+* Install the [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+
+The created resource group and the function app name should be provided to the placeholders shown in the above generated usage instructions from the compiler.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:65]:
```suggestion
The created resource group and the function app name should be provided to the placeholders shown in the above-generated usage instructions from the compiler. 
```

File: learn/deployment/azurefunctions.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+In order to deploy a Ballerina function in Azure Functions, the following prerequisites must be met.
+
+* Create an Azure [Function App](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal) with the given resource group with the following requirements:
+   - Runtime stack - `Java`
+   - Hosting operating system - `Windows` (default; Linux is not supported in Azure for custom handlers at the moment)
+* Install the [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+
+The created resource group and the function app name should be provided to the placeholders shown in the above generated usage instructions from the compiler. 
+
+A custom [`host.json`](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) file for the Azure Functions deployment can be optionally provided by placing a `host.json` file in the current working directory where the Ballerina build is done. The required `host.json` properties are provided/overridden by the values derived from the source code by the compiler extension.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:67]:
```suggestion
A custom [`host.json`](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) file for the Azure Functions deployment can be optionally provided by placing a `host.json` file in the current working directory in which the Ballerina build is done. The required `host.json` properties are provided/overridden by the values derived from the source code by the compiler extension. 
```

File: learn/deployment/azurefunctions.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+* Create an Azure [Function App](https://docs.microsoft.com/en-us/azure/azure-functions/functions-create-function-app-portal) with the given resource group with the following requirements:
+   - Runtime stack - `Java`
+   - Hosting operating system - `Windows` (default; Linux is not supported in Azure for custom handlers at the moment)
+* Install the [Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
+
+The created resource group and the function app name should be provided to the placeholders shown in the above generated usage instructions from the compiler. 
+
+A custom [`host.json`](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) file for the Azure Functions deployment can be optionally provided by placing a `host.json` file in the current working directory where the Ballerina build is done. The required `host.json` properties are provided/overridden by the values derived from the source code by the compiler extension. 
+
+A sample execution to deploy the functions to Azure Functions is shown below.
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:69]:
```suggestion
A sample execution to deploy the functions to Azure functions is shown below. 
```

File: learn/deployment/azurefunctions.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+  "received_time": "2020-07-15T07:32:21.9780071Z",
+  "site_name": "functions1777",
+  "start_time": "2020-07-15T07:32:23.2044517Z",
+  "status": 4,
+  "status_text": "",
+  "url": "https://functions1777.scm.azurewebsites.net/api/deployments/latest"
+}
+```
+
+The deployed Azure Function can be tested by invoking it using an HTTP client, such as CURL:
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:100]:
```suggestion
The deployed Azure Function can be tested by invoking it using an HTTP client such as CURL:
```

File: learn/deployment/azurefunctions.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+A full sample with all the supported Azure Functions triggers and bindings in Ballerina can be found [here](https://github.com/ballerina-platform/module-ballerinax-azure.functions/blob/master/azurefunctions-extension-examples/examples/azure-functions-deployment/azure_functions_deployment.bal).
```
Comment (praneesha) [inline learn/deployment/azurefunctions.md:107]:
```suggestion
For a full sample with all the supported Azure functions triggers and bindings in Ballerina, see [Azure Functions Deployment Example](https://github.com/ballerina-platform/module-ballerinax-azure.functions/blob/master/azurefunctions-extension-examples/examples/azure-functions-deployment/azure_functions_deployment.bal).
```

File: learn/deployment/azurefunctions.md (hunk 4 above)
Comment (lafernando) [reply learn/deployment/azurefunctions.md:14]: "Azure Function" is the actual service/entity, that is why it was capitalized.

File: learn/deployment/azurefunctions.md (hunk 6 above)
Comment (lafernando) [reply learn/deployment/azurefunctions.md:18]: Same as above.

File: learn/deployment/azurefunctions.md (hunk 4 above)
Comment (lafernando) [reply learn/deployment/azurefunctions.md:14]: Actually, the product is "Azure Functions" .. I guess we can't say something as a "Azure Function". Let me rename that with something else. I can probably say, "The Ballerina Azure Functions extension provides the functionality to expose a Ballerina function as a serverless function in the Azure Functions platform."

File: learn/deployment/azurefunctions.md (hunk 11 above)
Comment (lafernando) [reply learn/deployment/azurefunctions.md:69]: Refers to the service, so it's correct.

File: learn/deployment/azurefunctions.md (hunk 4 above)
Comment (praneesha) [reply learn/deployment/azurefunctions.md]: +1

**Pull request merged**